File: typescript/index1.ts
type Person = {
    team:string,
    name:string,
    phone:string,
    age:number
}

const getPerson = (team:string, name:string, age:number, phone:string):Promise<Person> =>
    Promise.resolve( name )
    .then( validateName )
    .then(
        (isValidName:boolean) => {
            if(isValidName === false) {
                return Promise.reject(new Error(`${name} is not a valid name for a Person type.`))
            } else {
                return phone
            }
        }
     )
    .then( validatePhone )
    .then(
        (isValidPhone:boolean) => {
            if(isValidPhone === false) {
                return Promise.reject(new Error(`${phone} is not a valid phone number for a Person type.`))
            } else {
                return age
            }
        }
    )
    .then( validateAge )
    .then(
        (isValidAge:boolean) => {
            if(isValidAge === false) {
                return Promise.reject(new Error(`${age} is not a valid age for a Person type.`))
            } else {
                return { name, phone, age, team}
            }
        }
    )


const validateName = (name:string):boolean => {
    // can't be 0 characters
    if(name.length < 1) {
        return false
    }
    // can't be a bunch-o-blanks
    if(name.split("").every(letter => letter === '')) {
        return false
    }
    return true
}

const validatePhone = (phone:string):boolean => {
    // must b 9 characters
    if(phone.length < 9) {
        return false
    }

    // all phone numbers are an Array of ints
    const isOkPhoneCharacter = (character:string):boolean => {
        if(character === ')'
            || character === '('
            || character === '-') {
                return true
            }
        const int = parseInt(character)
        if(typeof int === 'number' && isNaN(int) === false) {
            return true
        } else {
            return false
        }
    }

    // assuming we remove whitespace, it's should be either digits or 3 symbols
    return phone.trim().split('').every(isOkPhoneCharacter)
}

const validateAge = (age:number):boolean => {
    if(isNaN(age) === false && age > -1) {
        return true
    } else {
        return false
    }
}

if (process.argv[1].split('/')[process.argv[1].split('/').length - 1] === import.meta.url.split('/')[import.meta.url.split('/').length - 1]) {
    // getPerson("Jesse", "804-555-1234")
    getPerson("red", "804-555-1234", 42, "Jesse")
    .then(console.log)
    .catch(console.log)
}

File: python/index1.py
from dataclasses import dataclass, astuple
from enum import Enum
from returns.result import Result, Success, Failure
from functools import reduce

class Team(Enum):
	Red = "red"
	Blue = "blue"

def team_to_str(teamName:Team) -> str:
	# match teamName:
	# 	case Team.Red:
	# 			return "red"
	# 	case Team.Blue:
	# 			return "blue"
	# 	case _:
	# 			return "uh"
	if teamName == Team.Red:
		return "red"
	elif teamName == Team.Blue:
		return "blue"
	else:
		return "uh"

@dataclass
class Person:
    team: Team
    name: str
    age: int
    phone: str
    
    def __str__(self):
        return f'Person(team={team_to_str(self.team)}, name={self.name}, age={self.age}, phone={self.phone})'

def get_person(team:Team, name:str, age:int, phone:str) -> Result[Person, str]:
	return (
		validate_name(name)
		.bind(lambda _: validate_phone(phone))
		.bind(lambda _: validate_age(age))
		.bind(lambda _: Success(Person(team, phone, age, name)))
	)

def all_blanks(acc, character):
	if acc == False:
		return False
	elif character == "":
		return True
	else:
		return False

def validate_name(name:str) -> Result[str, str]:
	# can't be 0 characters
	if len(name) < 1:
		return Failure('name cannot be blank')
	
	# can't be a bunch-o-blanks
	if reduce(all_blanks, name.split(), True) == True:
		return Failure('name cannot be a bunch of blanks')
	
	return Success(name)

def validate_phone(phone:str) -> Result[str, str]:
	if len(phone) < 9:
		return Failure('phone cannot be less than 9 characters')

	if reduce(is_ok_phone_character, list(phone.strip()), True) == False:
		return Failure('phone has an unknown character, it must be either a number or ), (, or - as we strip our blank spaces')
	
	return Success(phone)

def is_parseable_int(character:str) -> bool:
	try:
		parsed = int(character)
		return True
	except:
		return False

def is_ok_phone_character(acc:bool, character:str) -> bool:
	if acc == False:
		return False
	
	if character == ')' or character == '(' or character == '-':
		return True
	
	return is_parseable_int(character)

def validate_age(age:int) -> Result[int, str]:
	if age < 0:
		return Failure('age must be greater than -1')

	return Success(age)

jesse = get_person(Team.Red, "804-555-1234", 42, "Jesse")
print(jesse)

